File: leetcode/problems_list/347-TopKFrequentElements/solution.py
"""
:type nums: List[int]
:type k: int
:rtype: List[int]
"""


class Solution(object):
    def topKFrequent(self, nums, k):
        top_elements = {}
        for num in nums:
            if top_elements.get(num) == None:
                top_elements[num] = 0
            top_elements[num] += 1
        return sorted(top_elements, key=top_elements.get, reverse=True)[:k]

File: leetcode/problems_list/347-TopKFrequentElements/solution.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
// Time Complexity: O(nlogn)
// Space Complexity: O(n)
var topKFrequent = function (nums, k) {
  const topElements = new Map();
  for (let num of nums) {
    if (!topElements.has(num)) topElements.set(num, 0);
    topElements.set(num, topElements.get(num) + 1);
  }
  const sortedElements = new Map(
    [...topElements.entries()].sort((a, b) => b[1] - a[1])
  );
  console.log(sortedElements.entries(), sortedElements.keys());
  return Array.from(sortedElements.keys()).slice(0, k);
};

/**
 * Solution using Object
 * Time Complexity: O(nlogn)
 * Space Complexity: O(n)
 * var topKFrequent = function(nums, k) {
 * const topElements = {}
 * for(let num of nums){
 *   topElements[num] = topElements[num] ? topElements[num]+1 : 1
 * }
 * return Object.keys(topElements).sort((a, b) => topElements[b]-topElements[a]).slice(0, k)
 * };
 */
